### ISSUE TITLE
Extend device support.

### ISSUE THREAD

**Issue by ambitious-octopus** [ultralytics/ultralytics#16930]
### Search before asking

- [X] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

We need to extend the list of support devices. At the moment we support just torch. devices, including CPU, cuda, and mps. We want to extend this to new devices like: tpu, intel npu, intel gpus and many more. We probably need to create a new ultralytics device class. 

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by Skillnoob** [comment]
I have a draft PR https://github.com/ultralytics/ultralytics/pull/15919 that adds support for any device that supports OpenCL (IGPU's, APU's & AMD GPU's) in the works. Training and inference works, but a lot of operations still fall back to the cpu since the pytorch_ocl package is not yet finished. More info can be found in that pr

**Comment by Y-T-G** [comment]
XPU support in PyTorch 2.5+ is much simpler.
https://pytorch.org/blog/intel-gpu-support-pytorch-2-5/

You only need to replace all `cuda` mentions in code with `xpu` to get it working.

```
!find "$(pip show ultralytics | awk '/Location: /{print $2}')/ultralytics" -type f -name "*.py" -exec sed -i 's/\bcuda\b/xpu/g' {} +
!find "$(pip show ultralytics | awk '/Location: /{print $2}')/ultralytics" -type f -name "*.py" -exec sed -i 's/\bCUDA\b/XPU/g' {} +
```

Someone on our Discord was able to successfully train with `ultralytics` by just doing that.

**Comment by Burhan-Q** [comment]
IIRC, the locations where a device is specified will have to get updated to use some form of `torch.device("DEVICE")` [[ref]](https://pytorch.org/docs/main/tensor_attributes.html#torch-device) including AMP, which is [experimental for XPU](https://pytorch.org/docs/main/amp.html#xpu-op-specific-behavior-experimental) at the moment.

**Comment by glenn-jocher** [comment]
Yes, you're correct. Looking at the reference material, we would need to update the `autocast` and `select_device` functions in `ultralytics/utils/torch_utils.py` to properly handle new device types. The current implementation is CUDA-specific, particularly in areas like autocast where it defaults to 'cuda' device. We should modify these to use a more generic `torch.device()` approach to support XPU, TPU and other accelerators. For XPU specifically, we'll need to be cautious with AMP support since it's still experimental in PyTorch.

**Comment by human-01** [comment]
### Working XPU training on Ultralytics 8.4.14 — specific fixes identified

I got YOLOv8n training working on **Intel Arc A770** (2× 16GB) with ultralytics 8.4.14, torch 2.10.0+xpu, Python 3.12 on Windows. Training completes successfully with correct GPU memory reporting (~4.2G reported). 100 epochs on a 225-image dataset in ~23 minutes.

#### Environment
- ultralytics 8.4.14
- torch 2.10.0+xpu (`pip install torch --index-url https://download.pytorch.org/whl/xpu`)
- Intel Arc A770 (16GB) × 2
- Windows 11, Python 3.12

> **Important:** `pip install ultralytics` pulls CPU-only `torch` as a dependency, overwriting the XPU build. You must reinstall XPU torch *after* installing/upgrading ultralytics:
> ```
> pip install ultralytics
> pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/xpu
> ```

#### 3 critical fixes (training blocked without these)

**1. `select_device()` rejects `"xpu"`** — `utils/torch_utils.py:185-208`

The `elif device:` branch assumes anything that isn't `"cpu"` or `"mps"` must be a CUDA device index. It sets `CUDA_VISIBLE_DEVICES="xpu"` and then raises `ValueError("Invalid CUDA 'device=xpu'")`.

Fix — add an XPU branch parallel to the `cpu`/`mps` checks (after line 182), or following @glenn-jocher's suggestion of a generic `torch.device()` approach:

```python
xpu = device == "xpu" or device.startswith("xpu:")
if xpu:
    if not (hasattr(torch, "xpu") and torch.xpu.is_available()):
        raise ValueError("XPU device requested but torch.xpu is not available")
    s += f"XPU ({torch.xpu.get_device_name(0)})\n"
    arg = "xpu"
```

Note: `select_device` is imported directly into `engine/trainer.py:59`, `engine/validator.py:40`, `engine/predictor.py:55`, and `engine/exporter.py:141` — they all hold their own reference, so a source fix here automatically covers all four.

**2. `GradScaler("cuda")` hard-coded** — `engine/trainer.py:333-335`

`_setup_train()` creates `torch.amp.GradScaler("cuda", enabled=self.amp)`. This crashes on XPU-only torch even when `amp=False` because the `"cuda"` backend string is validated at construction time.

Fix — use `self.device.type` instead of the hard-coded `"cuda"`:

```python
self.scaler = (
    torch.amp.GradScaler(self.device.type, enabled=self.amp) if TORCH_2_4
    else torch.cuda.amp.GradScaler(enabled=self.amp)
)
```

**3. `_get_memory()` falls through to CUDA** — `engine/trainer.py:580-591`

The `elif self.device.type != "cpu"` branch catches XPU devices but calls `torch.cuda.memory_reserved()` and `torch.cuda.get_device_properties()`, which return 0 or fail.

Fix — add an XPU branch before the CUDA fallthrough:

```python
elif self.device.type == "xpu":
    memory = torch.xpu.memory_reserved()
    if fraction:
        total = torch.xpu.get_device_properties(self.device).total_memory
```

#### 4 additional issues (training works without fixing these, but should be addressed)

4. **`_clear_memory()`** (`trainer.py:600-605`) — same pattern as `_get_memory()`, calls `torch.cuda.empty_cache()` for all non-CPU/non-MPS devices. Should add `torch.xpu.empty_cache()` for XPU.
5. **`autocast()` defaults to `"cuda"`** (`torch_utils.py:87-112`) — the trainer calls `autocast(self.amp)` at line 429 without passing a device arg, so it defaults to `"cuda"`. Should pass `self.device.type` through. AMP on XPU is experimental but functional in torch 2.10.
6. **`check_amp()` returns `False` for XPU** (`checks.py:848-849`) — only recognizes CUDA and MPS. XPU does support AMP via `torch.amp.autocast("xpu")`.
7. **OOM handler** (`trainer.py:446`) — catches `torch.cuda.OutOfMemoryError` only. XPU raises `torch.xpu.OutOfMemoryError` (or `RuntimeError`).

#### Device-generic approach

Per @glenn-jocher's comment, the ideal fix is making these device-generic rather than adding per-device branches. For example, `GradScaler(self.device.type)` and `autocast(device_type=self.device.type)` already work across CUDA, XPU, and MPS in PyTorch 2.4+. The memory functions could use a registry pattern:

```python
_MEMORY_FNS = {
    "cuda": (torch.cuda.memory_reserved, torch.cuda.get_device_properties, torch.cuda.empty_cache),
    "xpu": (torch.xpu.memory_reserved, torch.xpu.get_device_properties, torch.xpu.empty_cache),
    "mps": (torch.mps.driver_allocated_memory, None, torch.mps.empty_cache),
}
```

I'm currently using runtime monkey patches (wrapping the original functions without modifying ultralytics on disk) and training is completing successfully. Happy to submit a PR with the source-level fixes if that would be helpful.

**Comment by glenn-jocher** [comment]
Thanks for the detailed write-up—please go ahead and open a PR that makes XPU device handling first-class by (1) teaching `select_device()` to accept `device="xpu"`/`"xpu:0"` without treating it like a CUDA index and (2) removing hard-coded `"cuda"` in AMP setup by threading `self.device.type` into our `autocast(enabled, device=...)` helper and `torch.amp.GradScaler(device_type=...)`; the two key utilities are in the [torch_utils reference](https://docs.ultralytics.com/reference/utils/torch_utils/) and if you can include `yolo checks` output (or `ultralytics.utils.checks.collect_system_info()` if you suspect mixed installs) plus a tiny repro per the [minimum reproducible example guide](https://docs.ultralytics.com/help/minimum-reproducible-example/), it’ll be much easier to review/merge.
